### PR TITLE
security: gh security recs

### PR DESCRIPTION
**Pull request by hay-kot**
## What type of PR is this?

_(REQUIRED)_

- security
- bug

## What this PR does / why we need it:

Addresses several security concerns raised by external audit/review

1. All user HTTP requests now use a custom Transport class that applies a timeout and blocks known internal address spaces
2. Note that you can't simple whitelist specific IP's, you must resolve the domain name first and then check the address space of the resolved IP. This prevents attackers from creating a DNS entry that resolves to a local address space and using that url to attack a Mealie instance.

Also, note that these are mitigations, they are not 100% fixes. 

## Which issue(s) this PR fixes:

- Closes #3364

## Special notes for your reviewer:

I've tested locally and all the automated testing should pass. If possible, I would like another reviewer to pull down these changes and attempt to 

1. Scrape a few recipes from different domains
2. Use the Scrape Image functionality to pull external images into Mealie from a few different domains.

## Testing

See Above

### CODE REVIEW

**Review by boc-the-git — CHANGES_REQUESTED** [review]
The code changes all look good.
In testing the happy path, everything went well (i.e. I can still import recipes and images from legitimate internet sites.

In testing the negative path (trying to access things on my internal network), things came unstuck.

This first series of images is when accessing my Vikunja instance via the URL/reverse proxy (note I installed ping in the dev container to confirm what it's resolving to):
image
image
image

Then for a further test, I popped the IP/port in directly, avoiding the reverse proxy and any DNS resolution:
image

@hay-kot I'm interested in whether you can replicate this scenario.

I dug a bit deeper, and wanted to get some evidence of what was happening, which this code change and error message shows.. effectively the IP is `None`, rather than `192.168.x.y` as I'd expect.
image

**Comment by p0lycarpio** [comment]
Along the same lines as the `ALLOW_SIGNUP` variable which was set to `False`, I suggest setting all groups to private by default when they are created. This is an additional security measure and also when using Mealie for the first time the user would be redirected to the login/welcome page instead of the empty group.
What is your opinion ?

The change would be made in the [group_preferences.py](https://github.com/mealie-recipes/mealie/blob/7f596c653bdeaaee64efd5c6ec5f5f0a3108c2f4/mealie/schema/group/group_preferences.py#L9) schema. I don't know the impacts though.

**Comment by boc-the-git** [comment]
@p0lycarpio do you want to raise a feature request for that (or discuss it on Discord)?
We wouldn't likely tackle that under this PR.

That's more a _privacy issue_ than a _security issue_, if you take my meaning.
I don't have strong feelings on it one way or the other; other people might.

**Comment by hay-kot** [comment]
@boc-the-git should be good now. Thanks to some folks on discord it's 99% less bad too!

File: mealie/pkgs/safehttp/transport.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+        # validate the request is not attempting to connect to a local IP
+        # This is a security measure to prevent SSRF attacks
+
+        ip: ipaddress.IPv4Address | ipaddress.IPv6Address | None = None
+
+        netloc = request.url.netloc.decode()
+        if ":" in netloc:  # Either an IP, or a hostname:port combo
+            netloc_parts = netloc.split(":")
+            if len(netloc_parts) > 1:  # netloc of username:password@hostname:port not supported
```
Comment (boc-the-git) [inline mealie/pkgs/safehttp/transport.py]: The check here doesn't align with the comment.. should this be checking for `... > 2`?
I won't stop this delaying a merge, but will circle back to it, via Discord.

File: mealie/pkgs/safehttp/transport.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+                raise InvalidDomainError(f"unsupported request format: {request.url} -> {ip}")
```
Comment (boc-the-git) [inline mealie/pkgs/safehttp/transport.py]: The `ip` variable has never been set. Assume that's unintended.

File: mealie/pkgs/safehttp/transport.py (hunk 1 above)
Comment (boc-the-git) [reply mealie/pkgs/safehttp/transport.py]: `... > 1` will always be true, except in some weird scenario where `:` was the first or last character of `netloc` (and tbh I'm not certain what happens in that scenario)

File: mealie/pkgs/safehttp/transport.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+            netloc = netloc_parts[0]
+
+            try:
+                ip = ipaddress.ip_address(netloc)
+            except ValueError:
+                if self._log:
+                    self._log.debug(f"failed to parse ip for {netloc=} falling back to domain resolution")
```
Comment (boc-the-git) [inline mealie/pkgs/safehttp/transport.py:56]: I'm not familiar with how `{netloc=}` is being used here.. assuming the `=` is a typo

Comment (boc-the-git) [reply mealie/pkgs/safehttp/transport.py:56]: From python.org..
image

File: mealie/pkgs/safehttp/transport.py [hunk 4]
```diff
@@ -46,8 +46,6 @@ async def handle_async_request(self, request):
         netloc = request.url.netloc.decode()
         if ":" in netloc:  # Either an IP, or a hostname:port combo
             netloc_parts = netloc.split(":")
-            if len(netloc_parts) > 1:  # netloc of username:password@hostname:port not supported
```
Comment (boc-the-git) [inline mealie/pkgs/safehttp/transport.py]: @hay-kot see this log.. showing that even when `user:pass` is included in the url, it was being stripped out before getting to this netloc variable. Thus, the check on parts is irrelevant here.

image

**Review by boc-the-git — APPROVED** [review]

**Comment by boc-the-git** [comment]
@hay-kot this has my blessing, but note I had to make a fix. Please review, and merge/release as desired.

File: mealie/pkgs/safehttp/transport.py (hunk 1 above)
Comment (tba-code) [reply mealie/pkgs/safehttp/transport.py]: It was a typo.

**Pull request merged**